cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -798,7 +864,7 @@ class AzureFileSystem::Impl {
 
   std::unique_ptr<DataLake::DataLakeServiceClient> datalake_service_client_;
   std::unique_ptr<Blobs::BlobServiceClient> blob_service_client_;
-  internal::HierarchicalNamespaceDetector hns_detector_;
+  HNSSupport cached_hns_support_ = HNSSupport::kUnknown;
```

Comment:
This split separates the stateful part of HNS checking from the request and error handling logic. To counter the possibility of inadvertent misuse of the value, I renamed it to `cached_...`. There are valid uses for this value directly and the name describes that it's a cached value that could contain much more than the two states of a boolean.

Reply:
> There are valid uses for this value directly

Can you give some examples? I can't think of any

Reply:
An operation could decide to use the `DataLakeFileSystemClient` if `cached_hns_support` in `kUnknown/kContainerNotFound/kEnabled` and set the value of `cached_hns_support_` based on the error/success handling of that operation. This would save the mandatory extra request in the uncached case.

Another scenario is if we were to add threads to the mix, we would like to avoid having multiple HNS check requests going in parallel by having `cached_hns_support_` be some kind of atomic variable or protected by a mutex that also protects other member variables in the `AzureFileSystem::Impl` class.

Reply:
To be honest I'm not really convinced by either of these, but it probably doesn't matter. 

Technically you could avoid the call to check for hierarchical NS but I don't think it really makes sense because of the very small performance cost and the high added complexity cost. `DataLakeFileSystemClient` largely works on flat NS accounts so it would require extra error handling on every call and as we know from the hierarchical namespace detection code Azure often gives strange responses and its difficult to determine if the failure is genuine or if the error is just because hierarchical namespace is detected. 

I don't really see why we would need to protects other member variables of `AzureFileSystem::Impl` behind the same mutex. I would have kept the separate detector class which could have just one mutex. When `AzureFileSystem::Impl` needs to check whether hierarchical NS is enabled in parallel it would hit the mutex but otherwise I would want `AzureFileSystem::Impl` to be free to make other calls to blob storage without any mutexs getting in the way. Additionally, I was under the impression that `AzureFileSystem` would never be used in multiple threads. I think only `RandomAccessFile::ReadAt` is used concurrently.

Reply:
I will consider extracting the memoized `HierarchicalNamespaceSupport()` check to a separate class (added to my TODO list together with splitting the `azurefs.cc` file), but I maintain that having an underlying function that doesn't do any mutable state manipulation (just the request to the backend) makes it easier to reason about the correctness and cost of the calls.

> AzureFileSystem would never be used in multiple threads. I think only RandomAccessFile::ReadAt is used concurrently.

I meant `AzureFileSystem` itself would be using multiple threads to run the operations.

Reply:
> ...its difficult to determine if the failure is genuine or if the error is just because hierarchical namespace is detected.

I noticed this and I'm working on changes that puts us in a position where we never have to make that distinction. Because we can never cover all possible cases and Azurite being very broken when we make Data Lake Storage API calls to it makes it even worse.